lib/Optimizer/Transforms/ResetBeforeReuse.cpp

```diff
@@ -0,0 +1,242 @@
+};
+
+class ResetAfterMeasurePattern : public OpRewritePattern<quake::MzOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  explicit ResetAfterMeasurePattern(MLIRContext *ctx, RegUseTracker &tracker)
+      : OpRewritePattern(ctx), tracker(tracker) {}
+
+  LogicalResult matchAndRewrite(quake::MzOp mz,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<Operation *> useOps;
+    for (Value measuredQubit : mz.getTargets()) {
+      auto *nextOp = getNextUse(measuredQubit, mz);
+      if (nextOp) {
+        // If the user is a reset op, nothing to do.
+        if (isa<quake::ResetOp>(nextOp)) {
+          continue;
+        }
+
+        // If this is a dealloc op, nothing to do.
+        if (isa<quake::DeallocOp>(nextOp)) {
+          continue;
+        }
+
+        // Insert reset
+        Location loc = mz->getLoc();
+        rewriter.setInsertionPointAfter(mz);
+        rewriter.create<quake::ResetOp>(loc, TypeRange{}, measuredQubit);
+        // Insert a conditional X to initialize qubit after reset.
+        auto measOut = mz.getMeasOut();
+        mlir::Value measBit = [&]() {
+          for (auto *out : measOut.getUsers()) {
+            // A mz may be accompanied by a store op, find that op.
+            if (auto disc = dyn_cast_if_present<quake::DiscriminateOp>(out)) {
+              rewriter.setInsertionPointAfter(disc);
+              return disc.getResult();
+            }
+          }
+          // No discriminate exists - create the discriminate Op
+          auto discOp = rewriter.create<quake::DiscriminateOp>(
+              loc, rewriter.getI1Type(), measOut);
+          return discOp.getResult();
+        }();
+        rewriter.create<cudaq::cc::IfOp>(
+            loc, TypeRange{}, measBit,
+            [&](OpBuilder &opBuilder, Location location, Region &region) {
+              region.push_back(new Block{});
+              auto &bodyBlock = region.front();
+              OpBuilder::InsertionGuard guad(opBuilder);
+              opBuilder.setInsertionPointToStart(&bodyBlock);
+              opBuilder.create<quake::XOp>(location, measuredQubit);
+              opBuilder.create<cudaq::cc::ContinueOp>(location);
+            });
+      } else {
+        LLVM_DEBUG(llvm::dbgs() << "No next use\n");
+      }
+    }
+
+    return failure();
+  }
+
+private:
+  Operation *getNextUse(Value qubit, Operation *op) const {
+    auto &dom = tracker.getDominanceInfo();
+    {
+      // Check direct use
+      const auto orderedUsers = sortUsers(qubit.getUsers(), dom);
+      assert(orderedUsers.size() > 0);
+      for (auto v : llvm::enumerate(orderedUsers)) {
+        if (v.value() == op && v.index() < (orderedUsers.size() - 1) &&
+            dom.dominates(op, orderedUsers[v.index() + 1])) {
+          return orderedUsers[v.index() + 1];
+        }
+      }
+    }
+
+    // No next use is found, check if this is an extracted qubit.
+    if (isa<quake::RefType>(qubit.getType())) {
```

Comment:
In reference semantics, you will have to account for every possible case of aliasing. References can alias.

So it's not as simple as looking at alloca and extract_ref. Correctness here means you have to consider calls, block arguments, various arithmetic operations, dynamic qubit refreences, and so on.

On the other hand, value semantics has already been built and makes all of this much, much easier.